### DISCUSSION TITLE
Font combinations

### DISCUSSION THREAD

**Discussion by bytex0a** [olsak/OpTeX#196] [Q&A] [answered]
I would like to combine two fonts as described in the luaotfload manual (p. 23), i.e. use a base font and replace some glyphs by those of a second font.

    \def \feats {-tlig;-liga;mode=base;-kern} 
    \def \fileone {EBGaramond12-Italic.otf}
    \def \filetwo {FreeMonoBoldOblique.otf}
    \def \filethree {SourceSansPro-ExtraLight.otf} 
    \input luaotfload.sty
    \font \one = file:\fileone :\feats
    \font \two = file:\filetwo :\feats
    \font \three = file:\filethree :\feats
    \font \onetwothree = "combo: 1 -> \fontid \one ;
    2 -> \fontid \two , 0x30-0x39;
    3 -> \fontid \three , 0x21*0x3f; "
    {\onetwothree \TeX —0123456789—?!} 
    \bye

Could this be achieved using OpTeX?

**Comment by olsak** [comment]
I replaced `\input luaotfload.sty` by `\initunifonts` in your example (because this OpTeX macro initializes luaotfload) and the code can be processed by OpTeX:

```
\def \feats {-tlig;-liga;mode=base;-kern} 
\def \fileone {EBGaramond-Italic.otf}
\def \filetwo {FreeMonoBoldOblique.otf}
\def \filethree {SourceSansPro-ExtraLight.otf} 
\initunifonts
\font \one = file:\fileone :\feats
\font \two = file:\filetwo :\feats
\font \three = file:\filethree :\feats
\font \onetwothree = "combo: 1 -> \fontid \one ;
2 -> \fontid \two , 0x30-0x39;
3 -> \fontid \three , 0x21*0x3f; "
{\onetwothree \TeX —0123456789—?!} 
\bye
``` 
Where is a problem?

**Comment by bytex0a** [comment]
I'd like to use such a combined font within the OpTeX font selection system. For example selecting font family Baskerville the lowercase glyphs a-z of the regular font should be taken from e.g. Times regular font (doesn't make sense, just for example).

22.10.2024 18:40:39 Petr Olšák ***@***.***>:

**Comment by olsak** [comment, marked as answer]
OK. Next example combines letters from Pagella-Italic and Heros-Bold. Moreover, the macro `\combofont` respects the current size of fonts given by the OpTeX font selection system.
```
\fontfam[Heros]
\fontfam[Pagella]

\fontdef\one{\Pagella\it}   % Pagella-Italic
\fontdef\two{\Heros\bf}     % Heros-Bold

\def\combofont{%
   \fontlet\one=\one \_sizespec \fontlet\two=\two \_sizespec % resize \one and \two to the current font size
   \font \combospec = "combo: 1 -> \fontid \one ; 2 -> \fontid \two , 0x30-0x39;"
   \combospec
}

\combofont Text 012345.
\medskip
\setfontsize{at14pt}\combofont Larger text 12345.

\bye
```

**Comment by bytex0a** [comment]
Many thanks! OpTeX is great.

22.10.2024 20:10:46 Petr Olšák ***@***.***>:
